src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,384 @@ public class SqlMutationEngine : IMutationEngine
+            if (operationType is EntityActionOperation.Insert && !string.IsNullOrEmpty(primaryKeyRoute))
+            {
+                locationHeaderURL = UriHelper.BuildAbsolute(
+                                        scheme: httpContext.Request.Scheme,
+                                        host: httpContext.Request.Host,
+                                        pathBase: GetBaseRouteFromConfig(),
+                                        path: httpContext.Request.Path);
+
+                locationHeaderURL = locationHeaderURL.EndsWith('/') ? locationHeaderURL + primaryKeyRoute : locationHeaderURL + "/" + primaryKeyRoute;
+            }
+
+            // When the database policy is defined for the read action, a select query in another roundtrip to the database was executed to fetch the results.
+            // So, the response of that database query is used to construct the final response to be returned.
+            if (isDatabasePolicyDefinedForReadAction)
+            {
+                return (jsonDocument is not null) ? new CreatedResult(location: locationHeaderURL, OkMutationResponse(jsonDocument.RootElement.Clone()).Value)
+                                                  : new CreatedResult(location: locationHeaderURL, OkMutationResponse(JsonDocument.Parse("[]").RootElement.Clone()).Value);
+            }
+
+            // When no database policy is defined for the read action, the results from the upsert database operation is
+            // used to construct the final response.
+            // When no read permission is configured for the role, or all the fields are excluded
+            // an empty response is returned.
+            return (isReadPermissionConfiguredForRole && resultRow.Count > 0) ? new CreatedResult(location: locationHeaderURL, OkMutationResponse(resultRow).Value)
+                                                                              : new CreatedResult(location: locationHeaderURL, OkMutationResponse(JsonDocument.Parse("[]").RootElement.Clone()).Value);
+        }
+
+        /// <summary>
+        /// Constructs and returns a HTTP 200 Ok response.
+        /// The response is constructed using results of the upsert database operation when database policy is not defined for the read permission.
+        /// If database policy is defined, the results of the subsequent select statement is used for constructing the response.
+        /// </summary>
+        /// <param name="resultRow">Reuslt of the upsert database operation</param>
+        /// <param name="jsonDocument">Result of the select database operation</param>
+        /// <param name="isReadPermissionConfiguredForRole">Indicates whether read permissions is configured for the role</param>
+        /// <param name="isDatabasePolicyDefinedForReadAction">Indicates whether database policy is configured for read action</param>
+        private static OkObjectResult ConstructOkMutationResponse(
```

Comment:
Again this method makes more sense in `SqlResponseHelpers`.

Reply:
Moved to SqlResponseHelpers